### PR TITLE
NEXRAD Level2 structured reader 

### LINKED ISSUE #160: ENH: Add a NEXRAD Level2 Reader
### Description

Add a reader for NEXRAD Level2 data, based on the one currently in the Python ARM Radar Toolkit (Py-ART)

### PR DESCRIPTION
**Pull request by kmuehlbauer**
- [ ] Closes #160 
- [x] Tests added
- [x] Changes are documented in `history.md`

This is a rewrite of #147 using parts of the structured reader of the iris/sigmet reader. This also enhances the ability of reading much more metadata from the nexrad level 2 archive.

Note: this only implements MSG_31 as of now. MSG_1 need to be included after adding such file to open-radar-data repo.

There is also relevant code over in https://github.com/jthielen/xradar/tree/nexrad-level2 by @jthielen, which should be added to this structured reader in a follow-up PR.

### CODE REVIEW

**Comment by kmuehlbauer** [comment]
Finally linting is good. Had outdated version on my machine :grimacing:

**Comment by kmuehlbauer** [comment]
@mgrover1 

This only reads the absolutely necessary metadata to get a hunch on sweeps/moments. Everything else is read when first requested by xarray (so it's lazy too). The reader uses numpy.memmap (like the sigmet reader). Will push the fixture in a minute...

**Comment by mgrover1** [comment]
yes! next week's radar community meeting?

File: tests/io/test_nexrad_level2.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+    for k, v in fh.meta_header.items():
+        print(k, v)
+
+    print("MSG_5", fh.msg_5)
+
+    print(len(fh.data_header))
+
+
+def test_open_nexrad_level2_file_uncompressed():
+    fname_gz = "/home/kai/data/daten/daten/radar_all_over_the_world/NEXRAD/KATX20160601_090111_V06_gz"
```
Comment (mgrover1) [inline tests/io/test_nexrad_level2.py:349]: @kmuehlbauer - we should have one of these sample files over in open-radar-data?

Comment (kmuehlbauer) [reply tests/io/test_nexrad_level2.py:349]: @mgrover1 Yes, the nice thing is, we can have one with the old uncompressed data and the same file with the bzip2 compressed records. I'll add this next week.

And sorry for breaking the PR here, I'm currently testing with the compression. Seems like I'll need to change the class inheritance, but overall it seems to work.

Comment (kmuehlbauer) [reply tests/io/test_nexrad_level2.py]: done

**Review by mgrover1 — APPROVED** [review]
This looks great - thanks for your hard work here @kmuehlbauer

**Comment by mgrover1** [comment]
The docs and test coverage look great - we can continue to iterate + test through other PRs - what do you think about merging @kmuehlbauer ?

**Comment by kmuehlbauer** [comment]
Let's wait after the meeting today. There are some issues which I want to show before we get this in.

**Comment by mgrover1** [comment]
I am putting together a notebook right now comparing the data fields from Py-ART + xradar here as well... which should help us understand some of the differences?

**Comment by kmuehlbauer** [comment]
> I am putting together a notebook right now comparing the data fields from Py-ART + xradar here as well... which should help us understand some of the differences?

Great idea! One major difference is that this PR does not mask the data. We would need to think how we want to handle masking in the future. But we can make first make a list with the current differences.

**Comment by kmuehlbauer** [comment]
@mgrover1 If you think we should move forward from here, please go ahead and merge. I'll add several issues dedicated to different aspects/downsides of this implementation. Please also add any issues you find to the tracker, so that we can work on fixing enhancing.

**Comment by mgrover1** [comment]
Great!! Agreed 😄

**Comment by mgrover1** [comment]
@kmuehlbauer - should we close out #40 here and open new issues related to more specific improvements?

**Comment by kmuehlbauer** [comment]
Or let's use this issue (#40) as a tracking issue. We could list all related PR/issues there? Can also pin that issue to the top. WDYT? I'm also good with closing #40.

**Comment by mgrover1** [comment]
Ohh I like that idea... that can be the issue we refer back to + scaffold out. We can rename it to NEXRAD Support, and refer to the different readers and improvements.

**Comment by mgrover1** [comment]
I created #160 to reflect this single reader, and modified #40 accordingly. Sound reasonable @kmuehlbauer ?

**Pull request merged**